lib/client/reexec/reexec.go

```diff
@@ -0,0 +1,154 @@
+// Teleport
+// Copyright (C) 2025 Gravitational, Inc.
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package reexec
+
+import (
+	"context"
+	"io"
+	"os"
+	"os/exec"
+	"os/signal"
+	"strings"
+	"time"
+
+	"github.com/gravitational/trace"
+)
+
+// notifyFunc signals on the returned channel when the passed function returns
+// true, or when the context is canceled.
+func notifyFunc(ctx context.Context, check func() bool) <-chan error {
+	errorCh := make(chan error, 1)
+	go func() {
+		for {
+			if check() {
+				errorCh <- nil
+				return
+			}
+			select {
+			case <-time.After(time.Second):
+			case <-ctx.Done():
+				errorCh <- ctx.Err()
+				return
+			}
+		}
+	}()
+	return errorCh
+}
+
+// notifySignal signals on the returned channel when the passed OS signal is
+// received, or when the context is canceled.
+func notifySignal(ctx context.Context, sig os.Signal) <-chan error {
+	errorCh := make(chan error, 1)
+	go func() {
+		sigCh := make(chan os.Signal, 1)
+		signal.Notify(sigCh, sig)
+		defer signal.Stop(sigCh)
+
+		select {
+		case <-sigCh:
+			errorCh <- nil
+		case <-ctx.Done():
+			errorCh <- ctx.Err()
+		}
+	}()
+	return errorCh
+}
+
+// notifyChildReady signals on the returned channel when the child process is
+// ready to be disowned.
+func notifyChildReady(ctx context.Context) <-chan error {
+	return notifySignal(ctx, childReadySignal)
+}
+
+// SignalChildReady signals to the parent process that this process is ready
+// to be disowned.
+func SignalChildReady(ppid int) error {
+	proc, err := os.FindProcess(ppid)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	return trace.Wrap(proc.Signal(childReadySignal))
```

Comment:
This is racy, the pid of the parent could've been reused by the time we hit this.